### PR TITLE
cgroup namespaces: ignore the mount.Root if we have cgroup namespaces

### PR DESCRIPTION
**Pull request by hallyn**
In a cgroup namespace, you can mount cgroupfs, and your namespace
root (say /docker1) becomes the root of the cgroup filesystem.  This
shows up as field 3 in the mountinfo.  This is unfortunately
ambiguous with a cgroupfs bind mount, and in this case we cannot use
that root as a prefix for our cgroup, since as far as we are concerned
our cgroup is '/', not '/docker1'.

So if cgroup namespaces are enabled (/proc/$$/ns/cgroup exists), then
assume that we haven't done any silly cgroupfs bind mount trickery,
and ignore the fs root field.

Signed-off-by: Serge Hallyn <serge.hallyn@ubuntu.com>

### CODE REVIEW

**Comment by mrunalp** [comment]
LGTM

**Comment by mrunalp** [comment]
@hallyn Are cgroups namespaces expected to be in 4.6? (BTW, thanks for your upstream work in reviving the patchset).

**Comment by hallyn** [comment]
@mrunalp That's the hope :)  They're in linux-next at the moment.

**Comment by mrunalp** [comment]
Awesome :) Looking forward to it!

**Comment by mlaventure** [comment]
Shouldn't this also check that cgroups ns was enabled for that container? Or does this work even in the initial cgroup namespace?

**Comment by hallyn** [comment]
How I understand this code to work is,

if the cgroupfs mount is a bind mount into a container, and the host's

/sys/fs/cgroup/freezer/lxc/c1

was bind mounted into the container such that

/sys/fs/cgroup/freezer in the container is a tmpfs, 'lxc/c1' are directories in
the tmpfs, and the bind mount is onto /sys/fs/cgroup/freezer/lxc/c1, then the
container will find that it is in cgroup '/lxc/c1', so to create /lxc/c1/x1,
it must create 'x1' under the mountpoint. So it concatenates the mountpoint,
/sys/fs/cgroup/freezer/lxc/c1, with the mount prefix removed from the wanted
path (/lxc/c1/x1 - /lxc/c1 = x1).

Simply put, when cgroup namespaces are enabled, this gimmick is not needed at
all because the container was able to mount a real cgroupfs.

Now unfortunately there is no way to absolutely tell which case we have going.
In both cases the fs shows up as 'cgroup'.  In both cases it remains possible
that current's cgroup start with the mount prefix:  with cgroup namespaces,
I could be in absolute cgroup /lxc/c1/lxc/c1, namespaced /lxc/c1.  So in
this patch I simply assume that if cgroup namespaces are available, then they
will be used.

**Comment by hqhq** [comment]
>  I simply assume that if cgroup namespaces are available, then they will be used

I'm afraid we don't have this assumption now, namespaces are configurable, this assumption probably would break backward compatibility on new kernel, right?

**Comment by cyphar** [comment]
@hqhq I'm fairly namespace joining used to be kernel-version backwards compatible (we don't join namespaces which don't exist and only emit warnings). I'm not sure now (reading over the code, it doesn't look like that's the case anymore).

**Comment by hqhq** [comment]
@cyphar I'm not worry about the kernel, but concern we can't use old `configs.json` on kernel which support cgroup namespace, because old json file has no cgroup ns configed, but runC will just join it and change the cgroup root.
Maybe not a big deal because we are not 1.0 yet, just think we can have a better way to check if we should ignore `mount.Root`.

If we gonna do a follow up PR to fix that, this PR LGTM.

**Comment by mlaventure** [comment]
I don't think runC should automatically join namespaces that are not listed in the config. That would be quite un-intuitive and limit some valid use cases.

If I'm not mistaken, atm we only join the namespaces that are actually listed within the config file. Which means that this PR should probably be updated to also check that a cgroup ns was joined before changing the value of `m.Root`.

**Comment by mrunalp** [comment]
I think we should probably wait for cgroups namespaces to be available before we merge this and other required support to runc.

**Comment by cyphar** [comment]
@hqhq Sorry, when I said "kernel-version backwards compatible" I was referring to runc being compatible with older kernels. I'm not sure that this is the case anymore (reading the code for joining namespaces, it looks like we error out if we can't join any one of them).

**Comment by hallyn** [comment]
(closing this until upstream churn settles down)

**Comment by cyphar** [comment]
@hallyn So, cgroup namespaces have been merged (in a very different state to the current proposal here). We can reopen this once #781 and the dependent PRs all get merged.

**Comment by hallyn** [comment]
This PR should not be needed as we now have /proc/self/mountinfo's cgroup entries virtualized wrt the cgns the same as the /proc/self/cgroup entries.

So while I've not tested it, I expect it to "just work" now.

**Comment by SpComb** [comment]
It seems that this patch is still required on Ubuntu xenial's linux 4.4 kernel, with the parent nsroot in `/proc/self/mountinfo`. Ubuntu's own docker.io packages carry this patch and work, but the upstream docker-engine package fails, note the broken `/sys/fs/docker/...` cgroup path:

```
level=error msg="containerd: start container" error="oci runtime error: could not synchronise with container process: stat /sys/fs/docker/13bd9f844f91631c0459d5fabdad8b2e555d7765eefbc7158b4db428b118b008: no such file or directory
```

This (or some other related issue) causes `docker run` to fail with errors like: `docker: Error response from daemon: Container command 'bash' not found or does not exist..`

**Comment by hallyn** [comment]
the kernel to fix that should have just cleared xenial-proposed.

**Comment by SpComb** [comment]
Cool, do you have a linux-image-generic version number for the package with the fix? I've been testing with 4.4.0-22-generic. I'd be happy to test against the upstream docker-engine package with the fixed kernel.

**Comment by hallyn** [comment]
I guess I was wrong about it clearing -proposed.  The fix is in
4.4.0-25.44

https://launchpad.net/ubuntu/+source/linux/4.4.0-25.44

**Comment by SpComb** [comment]
I can confirm that with Ubuntu xenial-proposed linux 4.4.0-25-generic including the cgroup namespace updates, `/proc/self/mountinfo` now shows a path of / within the lxd cgroup namespace, and the unpatched upstream docker-engine 1.11.2 package works correctly.

So assuming it's the same behaviour as on linux 4.6, this should be fine now.

**Comment by hallyn** [comment]
Great - thx for testing.
